compiler/optimizer/demand.go

```diff
@@ -0,0 +1,198 @@
+package optimizer
+
+import (
+	//"fmt"
+	"github.com/brimdata/zed/compiler/ast/dag"
+)
+
+type DemandAll struct{}
+type DemandKeys map[string]Demand // No empty values.
+
+type Demand interface {
+	isDemand()
+}
+
+func (demand DemandAll) isDemand()  {}
+func (demand DemandKeys) isDemand() {}
+
+func demandNone() Demand {
+	return DemandKeys(make(map[string]Demand, 0))
+}
+
+func demandIsValid(demand Demand) bool {
+	switch demand := demand.(type) {
+	case DemandAll:
+		return true
+	case DemandKeys:
+		for _, v := range demand {
+			if !demandIsValid(v) || demandIsEmpty(v) {
+				return false
+			}
+		}
+		return true
+	default:
+		panic("Unreachable")
+	}
+}
+
+func demandIsEmpty(demand Demand) bool {
+	switch demand := demand.(type) {
+	case DemandAll:
+		return false
+	case DemandKeys:
+		return len(demand) == 0
+	default:
+		panic("Unreachable")
+	}
+}
+
+func demandKey(key string, value Demand) Demand {
+	if demandIsEmpty(value) {
+		return value
+	}
+	demand := DemandKeys(make(map[string]Demand, 1))
+	demand[key] = value
+	return demand
+}
+
+func demandUnion(a Demand, b Demand) Demand {
+	if _, ok := a.(DemandAll); ok {
+		return a
+	}
+	if _, ok := b.(DemandAll); ok {
+		return b
+	}
+
+	{
+		a := a.(DemandKeys)
+		b := b.(DemandKeys)
+
+		demand := DemandKeys(make(map[string]Demand, len(a)+len(b)))
+		for k, v := range a {
+			demand[k] = v
+		}
+		for k, v := range b {
+			if v2, ok := a[k]; ok {
+				demand[k] = demandUnion(v, v2)
+			} else {
+				demand[k] = v
+			}
+		}
+		return demand
+	}
```

Comment:
Nit: Indent this less.
```suggestion
	a, b := a.(DemandKeys), b.(DemandKeys)
	demand := DemandKeys(make(map[string]Demand, len(a)+len(b)))
	for k, v := range a {
		demand[k] = v
	}
	for k, v := range b {
		if v2, ok := a[k]; ok {
			demand[k] = demandUnion(v, v2)
		} else {
			demand[k] = v
		}
	}
	return demand
```

Reply:
This does not compile - `no new variables`.

Reply:
Ah, I forgot that `a := a.(DemandKeys)` requires a new scope. 🤦